### ISSUE TITLE
Deebot Ozmo 950 not updating state until restart of server

### ISSUE THREAD

**Issue by selmaohneh** [And3rsL/Deebot-for-Home-Assistant#178]
### Is there an existing issue for this?

- [X] I have searched the existing issues and no issue is describing my issue

### The problem

Looking at my vacuum entity from the developer tools it says it is in state `Docked`. 

After that I start a cleaning procedure by calling the service `vacuum.start`. My Osmo 950 starts just fine and cleans the house.

While cleaning, I look at the entity from the developer tools again, it still says: `Docked`. I can send other commands like `vacuum.stop`, `vacuum.pause`, etc.. They all work fine. Only the state is not updating.

I then restart my home assistant server while cleaning. After the restart the state updates as expected and shows `Cleaning`. Even when done the state changes to `Docked` as expected.

### On which deebot vacuum you have the issue?

Deebot Ozmo 950

### Which version of the Deebot component are you using?

3.1.0

### Which version of Home Assistant are you using?

2021.8.1

### What type of installation are you running?

Home Assistant OS

### Country

de

### Continent

eu

### Instance type

Cloud

**Comment by edenhaus** [comment]
We need your logs otherwise we cannot help you.
I think you have a problem with MQTT,. but there must be something in the logs

**Comment by selmaohneh** [comment]
And yes... there seems to be something wrong with my mosquitto:

1628674828: New connection from 172.30.32.2 on port 1883.
1628674828: Socket error on client <unknown>, disconnecting.
1628674948: New connection from 172.30.32.2 on port 1883.
1628674948: Socket error on client <unknown>, disconnecting.
1628675068: New connection from 172.30.32.2 on port 1883.
1628675068: Socket error on client <unknown>, disconnecting.
1628675188: New connection from 172.30.32.2 on port 1883.
1628675188: Socket error on client <unknown>, disconnecting.
1628675308: New connection from 172.30.32.2 on port 1883.

**Comment by capstan1** [comment]
Have the exact same issue with my T8+. The sensors (like cleaned area or cleaned time) go unknown when this happens.

As a workaround I'm using a automation calling the service homeassistant.reload_config_entry with my robot as entity id, when the state of the sensors changes to unknown or unavailable for 5 minutes. This restarts the integration and state and sensors are available afterwards.

**Comment by edenhaus** [comment]
Sorry guys, but if you want my help I need your logs from all loggers describe in de Readme.
The sensors only are set to unknown/unavailable if the api reports back that your robot is offline or there exists a connection issue.
Is your internet connection stable?

**Comment by capstan1** [comment]
> Sorry guys, but if you want my help I need your logs from all loggers describe in de Readme.
> The sensors only are set to unknown/unavailable if the api reports back that your robot is offline or there exists a connection issue.
> Is your internet connection stable?

Sure, I'll do. I had a restart since it occured last 3 days ago and before finding this issue. Just wanted to let OP know the workaround I found.

Will post here as soon as it appears the next time.

**Comment by capstan1** [comment]
So, it happened again, but with all variables "frozen". So, no "unavailable" or "unknown", the states just don't change when starting the robot. So it's hard to know, when the issue started. Only things I could find are these:

```
Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/deebot/hub.py", line 104, in _check_status_task
    await self._check_status_function()
  File "/config/custom_components/deebot/hub.py", line 107, in _check_status_function
    devices = await self._ecovacs_api.get_devices()
  File "/usr/local/lib/python3.9/site-packages/deebotozmo/ecovacs_api.py", line 100, in get_devices
    json = await self.__call_portal_api(self.API_APPSVR_APP, data)
  File "/usr/local/lib/python3.9/site-packages/deebotozmo/ecovacs_api.py", line 224, in __call_portal_api
    res.raise_for_status()
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1000, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 504, message='Gateway Time-out', url=URL('https://portal-eu.ecouser.net/api/appsvr/app.do')
```

```
Logger: deebotozmo.vacuum_bot
Source: /usr/local/lib/python3.9/site-packages/deebotozmo/vacuum_bot.py:158
First occurred: 13. August 2021, 16:46:52 (1 occurrences)
Last logged: 13. August 2021, 16:46:52

Invalid Event cleanlogs: {}
```

**Comment by selmaohneh** [comment]
> So, it happened again, but with all variables "frozen". So, no "unavailable" or "unknown", the states just don't change when starting the robot. So it's hard to know, when the issue started. Only things I could find are these:
> 
> ```
> Error doing job: Task exception was never retrieved
> Traceback (most recent call last):
>   File "/config/custom_components/deebot/hub.py", line 104, in _check_status_task
>     await self._check_status_function()
>   File "/config/custom_components/deebot/hub.py", line 107, in _check_status_function
>     devices = await self._ecovacs_api.get_devices()
>   File "/usr/local/lib/python3.9/site-packages/deebotozmo/ecovacs_api.py", line 100, in get_devices
>     json = await self.__call_portal_api(self.API_APPSVR_APP, data)
>   File "/usr/local/lib/python3.9/site-packages/deebotozmo/ecovacs_api.py", line 224, in __call_portal_api
>     res.raise_for_status()
>   File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1000, in raise_for_status
>     raise ClientResponseError(
> aiohttp.client_exceptions.ClientResponseError: 504, message='Gateway Time-out', url=URL('https://portal-eu.ecouser.net/api/appsvr/app.do')
> ```
> 
> ```
> Logger: deebotozmo.vacuum_bot
> Source: /usr/local/lib/python3.9/site-packages/deebotozmo/vacuum_bot.py:158
> First occurred: 13. August 2021, 16:46:52 (1 occurrences)
> Last logged: 13. August 2021, 16:46:52
> 
> Invalid Event cleanlogs: {}
> ```

Yep - just to give my +1: The states are *frozen*, i.e. the vacuum is still `docked` although the cleaning already started. The state is *not* `unavailable`. Only a restart or the workaround from @capstan1 (reloading the config entry before starting a cleaning session) solves the issue.

Still nothing obvious inside the logs :-/

**Comment by edenhaus** [comment]
`aiohttp.client_exceptions.ClientResponseError: 504, message='Gateway Time-out', url=URL('https://portal-eu.ecouser.net/api/appsvr/app.do')`

I don't know how much are you in in HTTP-Status codes, but all 500 codes mean there is a problem on the server side (ecovacs).
I will give it a look to improve the retry if some requests fails but at the end ecovacs must improve their server.

The invalid cleanlogs event is something else. As I don't use the cleanlogs it has a low priority for me. You must always remember that I developing this in my freetime

**Comment by Edo78** [comment]
> You must always remember that I developing this in my freetime
And we are really grateful for that.

BTW do you have a buymeacoffe account so I can send you a couple of bucks just as as a token of appreciation

**Comment by edenhaus** [comment]
> BTW do you have a buymeacoffe account so I can send you a couple of bucks just as as a token of appreciation

@Edo78 I don't have one, it's Andreas one. But for sure, he is happy for your support :)

I have created `3.1.1b1`, which should handle better the client exceptions... A warning will still be logged, but automatically it will reconnect each 60 seconds. Let me please know, if this solved the issue and also how often the connection will be lost. (Therefore the level warning for now)

As I use bumper (local server, which simulates the ecovacs server part) I don't have connection issues as HA and bumper run on the same machine.

**Comment by Yterz** [comment]
I had this problem with 3.1.1 too, commands sending but no update on any sensor/status a while after a HA restart. Last week I fully updated HA to 2021.9 and Deebot-for-HA to 3.1.3, and deactivated my daily-reboot-HA automation (I know, not the best patch...). Informations from my Deebot have been properly updated for 2 cleaning since that without needing a HA reboot.

I did got a Status="Error" once though, with last_error=103, then last_error=0 when HA requested a new cleaning (Status remained ="Error", then moved to "Returning to Dock" at the end of cleaning). But it seems to me it came from the Deebot itself, not HA or the integration (the map got tilted by the robot, even in the application).
What would be the best patch in HA for that, to get back the proper status of the Deebot ? An automation like When Status=Error for > xx minute, Do deebot.refresh - Part: Status ? Or Do homeassistant.reload_config_entry - entity_id: vacuum.MyDeebot ? Or it wouldn't change a thing and I need to call a stop + go dock or something ?

(Device: Deebot T8 AIVI+ on CN country ; Home Assistant 2021.9.6 ; Hardware: Raspberry Pi 4 ; Deebot-for-hassio : 3.1.3 ; )

**Comment by edenhaus** [comment]
For better understanding, the `last_error` is directly coming from the vacuum and also the rest to 0. I only set the status to error, when the code != 0. I can't reproduce this issue, so it's hard for me to give you more details on it.

The next step would be for sure to find out, what error 103 means. Probably we need to decompile and analise the Andorid App.

For your question. I think a service call on `deebot.refresh` is enough as it will request the actual status manually.
A reload of the config entry is probably not needed, but you should check your logs. Maybe we can findout more and I can fix this issue directly in the code for all.

**For all that have this issue. Please provide me with logs and information, what did you find out. I cannot reproduce this issue yet**

**Comment by Yterz** [comment]
After a few days without any HA reboot/refresh, status failed to be properly reported again, even with commande still properly sending. So I guess it was stabler but not fixed... :/

I guess a fix for now would be at the end of automations sending any command to add a `homeassistant.reload_config_entry` - `entity_id: vacuum.MyDeebot`. I didn't applied that yet and added instead the logger to my configuration.yaml. I'll try to leave my HA untouched for few days and report what I get if/when I encounter that again

> For better understanding, the `last_error` is directly coming from the vacuum and also the rest to 0. I only set the status to error, when the code != 0.

It match what I got, vacuum got stuck on my carpet when it got Status=Error and last_error=103 with an according notification on the application. The weird part was that when last_error went back to =0, Status remained =Error (I un-stuck the vacuum and send it back cleaning again). Status only changed by itself to GoingToBase when that new cleaning was done and it went home.
But I feel it was unrelated to that topic issue, and if I encounter it again I'll open a new issue with logs I'll try to get :)

**Comment by edenhaus** [comment]
Reloading the config entry is not a solution. it‘s only a temporary workaround. I think it‘s probably a connection issue. Normally the used mqtt library is reconnecting automatically, but maybe there exist a edge case where they don‘t reconnect.
Please check also your logs for stacktraces. Yesterday I solved a Timeout exception, which I forgotten to handle and this was logged by the asyncio logger.

> It match what I got, vacuum got stuck on my carpet when it got Status=Error and last_error=103 with an according notification on the application. The weird part was that when last_error went back to =0, Status remained =Error (I un-stuck the vacuum and send it back cleaning again). Status only changed by itself to GoingToBase when that new cleaning was done and it went home.
> But I feel it was unrelated to that topic issue, and if I encounter it again I'll open a new issue with logs I'll try to get :)

When the error is reseted to 0, I do nothing (meaning I don‘t change the state). The reason is that someone needs to analyse the messages from the robot further, so we can clearly identify the correct state. Probably the vacuum is in the idle state but we need to verify it.

**Comment by Edo78** [comment]
> @Edo78 @capstan1 @selmaohneh Can you please check if you have also the following warnings in your log (As it‘s a warming you don‘t need to change the level, as warnings are logged by default)? Is your country also set to cn?

Right now I have no warnings at all regarding this component and, yes, one of my accounts is set to `cn`

BTW I'm glad you finally opened a buymeacoffe and I already send you some deserved $$$ (or to be more accurate €€€)

**Comment by Edo78** [comment]
> Maybe I have found the real issue. When you did have the problem, was your instance running more then 7 days?
> 
> Today I found out, that the token are only valid for 7 days. It's still strange that mqtt is failing and the rest api not, but it could be that ecovacs checks only the token validity on mqtt.

This could explain why I experienced this issue only with an high uptime. Nice catch :+1:

**Comment by edenhaus** [comment]
With the next version the new authentication system is coming which should refresh automatically the tokens. As my HA instances is not running 7 days without restart it's hard to test it :D
With the new version I will close this issue as a lot will be changed about authentication. If the issue still exists, please open a new issue

**Comment by selmaohneh** [comment]
Still not working for me though :-/

![image](https://user-images.githubusercontent.com/29881939/142638540-44abdace-9e45-4a4f-a152-2cdab8605d5a.png)

![image](https://user-images.githubusercontent.com/29881939/142638591-3464dfd3-96c5-4d9f-8793-d53b84ff0697.png)
